core/src/main/java/me/mykindos/betterpvp/core/client/offlinemessages/OfflineMessagesRepository.java

```diff
@@ -0,0 +1,99 @@
+    private final Database database;
+
+    @Inject
+    public OfflineMessagesRepository(Database database) {
+        this.database = database;
+    }
+
+    @Override
+    public void save(OfflineMessage offlineMessage) {
+        String query = "INSERT INTO offline_messages (Client, Time, Action, Message) VALUES (?, ?, ?, ?)";
+        UtilServer.runTaskAsync(JavaPlugin.getPlugin(Core.class), () -> {
+            Statement statement = new Statement(query,
+                    new UuidStatementValue(offlineMessage.getClient()),
+                    new LongStatementValue(offlineMessage.getTime()),
+                    new StringStatementValue(offlineMessage.getAction().name()),
+                    new StringStatementValue(offlineMessage.getRawContent())
+            );
+
+            database.executeUpdate(statement, TargetDatabase.GLOBAL);
+            log.info("Saved offline message {} to database", offlineMessage).submit();
+        });
+    }
+
+    /**
+     * Gets the offline messages sent after last logout
+     * Must be called after client properties are loaded
+     * @param client the client to load the messages for
+     * @return the list of offline messages, sorted by most recent
+     */
+    public List<OfflineMessage> getNewOfflineMessagesForClient(Client client) {
+        return getOfflineMessagesForClient(client.getUniqueId(), (long) client.getProperty(ClientProperty.LAST_LOGIN).orElse(0));
+    }
+
+    /**
+     * Gets the offline messages sent after time
+     * @param clientID the id of the client to load the messages for
+     * @param time the start time to retrieve messages
+     * @return the list of offline messages, sorted by most recent
+     */
+    public List<OfflineMessage> getOfflineMessagesForClient(UUID clientID, long time) {
+        List<OfflineMessage> offlineMessages = new ArrayList<>();
+
+        String query = "CALL GetOfflineMessagesByTime(?, ?);";
+        Statement statement = new Statement(query,
+                new UuidStatementValue(clientID),
+                new LongStatementValue(time)
+        );
+
+        CachedRowSet result = database.executeQuery(statement, TargetDatabase.GLOBAL);
+        try {
+            while (result.next()) {
+
+
+                long messageTime = result.getLong(1);
+                String messageAction = result.getString(2);
+                String messageMessage = result.getString(3);
+
+                offlineMessages.add(new OfflineMessage(
+                        clientID,
+                        messageTime,
+                        OfflineMessage.Action.fromString(messageAction),
+                        messageMessage
+                        )
+                );
+            }
+        }
+        catch (SQLException e) {
+                log.error("Error while retrieving offline messages for client {}", clientID, e).submit();
+        }
+        return offlineMessages;
+    }
```

Comment:
My gut tells me this is running on the main thread, fetch with a completeablefuture and use whenComplete to do what you want with it

Reply:
whenComplete? why?

thenAcceptAsync is a much better alternative

Reply:
Why is it a much better alternative

Reply:
whenComplete also gives exceptions: which, if you're always handling it as successful is a bit dodgy (and since this is a db call, I'd assume you want some error logging).

thenAcceptAsync is given for success only, you handle failure with `exceptionally`

It gives much better handling, allows you to write code that makes it clear how you're handling exceptions.

This case isn't on the hot path, so I'd argue exceptionally isn't important, but assuming things are successfully ran isn't a good idea.